In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf

In [2]:
pip install tensorflow 

Note: you may need to restart the kernel to use updated packages.


In [3]:
pd.options.mode.chained_assignment = None  # default='warn'

In [4]:
# Download marketing dataset
df_marketing = pd.read_csv("C:/Users/vishn/Downloads/BostonHousing.csv")
df_marketing

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO  LSTAT  MEDV  CAT. MEDV  
0       15.3   4.98  24.0          0  
1       17.8   9.14  21.6          0  
2       17.8   4.03  34.7          1  
3       18.7   2.94  33.4          1  
4       18.7   5.33  36.2          1  
..       ...    ...   ...        ...  
501     21.0   9.67  22.4          0  
502     21.0   9.08  20.6          0  
503     21.0   5.64  23.9          0  
504     21.0   6.48  22.0          0  
505     21.0   7.88  11.9          0  

[506 rows x 14 columns]

In [5]:
# Make sure data types are correct (no objects)
df_marketing

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO  LSTAT  MEDV  CAT. MEDV  
0       15.3   4.98  24.0          0  
1       17.8   9.14  21.6          0  
2       17.8   4.03  34.7          1  
3       18.7   2.94  33.4          1  
4       18.7   5.33  36.2          1  
..       ...    ...   ...        ...  
501     21.0   9.67  22.4          0  
502     21.0   9.08  20.6          0  
503     21.0   5.64  23.9          0  
504     21.0   6.48  22.0          0  
505     21.0   7.88  11.9          0  

[506 rows x 14 columns]

In [6]:
# Pop label
target = df_marketing.pop('CAT. MEDV')

In [7]:
# Define features
marketing_feature_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'LSTAT', 'MEDV'] 
# Update with relevant feature names
marketing_features = df_marketing[marketing_feature_names]

In [8]:
# Convert data to Tensorflow tensor
marketing_features = tf.convert_to_tensor(marketing_features)
marketing_features

<tf.Tensor: shape=(506, 13), dtype=float64, numpy=
array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 4.9800e+00,
        2.4000e+01],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 9.1400e+00,
        2.1600e+01],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 4.0300e+00,
        3.4700e+01],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 5.6400e+00,
        2.3900e+01],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 6.4800e+00,
        2.2000e+01],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 7.8800e+00,
        1.1900e+01]])>

In [9]:
# Normalize data
normalizer = tf.keras.layers.Normalization(axis=-1)
normalizer.adapt(marketing_features)

In [10]:
# Define the model
def get_marketing_model():
  model = tf.keras.Sequential([
    normalizer,
    tf.keras.layers.Dense(10, activation='relu'),
    tf.keras.layers.Dense(10, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')  # Binary classification output
  ])

  model.compile(optimizer='adam',
                loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
                metrics=['accuracy'])
  return model


In [11]:
# Train model
BATCH_SIZE = 128
model = get_marketing_model()
model.fit(marketing_features, target, epochs=100, batch_size=BATCH_SIZE)


Epoch 1/100



C:\Users\vishn\anaconda3\Lib\site-packages\keras\src\backend.py:5818: UserWarning: "`binary_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Sigmoid activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(



4/4 [==============================] - 2s 5ms/step - loss: 0.7056 - accuracy: 0.5395
Epoch 2/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6906 - accuracy: 0.5593
Epoch 3/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6770 - accuracy: 0.5771
Epoch 4/100
4/4 [==============================] - 0s 3ms/step - loss: 0.6629 - accuracy: 0.5949
Epoch 5/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6499 - accuracy: 0.6206
Epoch 6/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6362 - accuracy: 0.6640
Epoch 7/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6230 - accuracy: 0.7213
Epoch 8/100
4/4 [==============================] - 0s 5ms/step - loss: 0.6094 - accuracy: 0.7787
Epoch 9/100
4/4 [==============================] - 0s 5ms/step - loss: 0.5951 - accuracy: 0.7964
Epoch 10/100
4/4 [==============================] - 0s 5ms/step - loss: 0.5807 - accuracy: 0.8182
Epoch 11/100
4/4 [======================

In [12]:
# Evaluate model
score = model.evaluate(marketing_features, target, verbose=1)
print(f'Test loss: {score[0]} / Test accuracy: {score[1]}')

16/16 [==============================] - 0s 4ms/step - loss: 0.0648 - accuracy: 0.9842
Test loss: 0.06483060121536255 / Test accuracy: 0.9841897487640381


In [20]:
# Predict labels for new records
df_new_records = pd.read_csv("C:/Users/vishn/Downloads/BostonHousing.csv")
new_records_features = df_new_records[marketing_feature_names]
new_predictions = model(new_records_features, training=False)

In [16]:
#display predictions
for i, prob in enumerate(new_predictions):
    label = 1 if prob[0] >= 0.5 else 0
    print(f"Example {i + 1} prediction: Purchase {'Made' if label == 1 else 'Not Made'} (Probability: {prob[0]:.2%})")

Example 1 prediction: Purchase Not Made (Probability: 16.80%)
Example 2 prediction: Purchase Not Made (Probability: 0.80%)
Example 3 prediction: Purchase Made (Probability: 69.54%)
Example 4 prediction: Purchase Made (Probability: 56.39%)
Example 5 prediction: Purchase Made (Probability: 72.82%)
Example 6 prediction: Purchase Not Made (Probability: 5.66%)
Example 7 prediction: Purchase Not Made (Probability: 0.87%)
Example 8 prediction: Purchase Not Made (Probability: 1.87%)
Example 9 prediction: Purchase Not Made (Probability: 0.02%)
Example 10 prediction: Purchase Not Made (Probability: 0.38%)
Example 11 prediction: Purchase Not Made (Probability: 0.51%)
Example 12 prediction: Purchase Not Made (Probability: 0.49%)
Example 13 prediction: Purchase Not Made (Probability: 0.31%)
Example 14 prediction: Purchase Not Made (Probability: 0.18%)
Example 15 prediction: Purchase Not Made (Probability: 0.19%)
Example 16 prediction: Purchase Not Made (Probability: 0.11%)
Example 17 prediction: Pu

# Enhancing Marketing Strategy Through Predictive Analytics: A Customer Purchase Prediction Model

1. Improved Decision-Making:
The model demonstrates a high accuracy rate, indicating its capability to make accurate predictions on whether a customer is likely to make a purchase. This can significantly enhance decision-making processes in marketing strategies.

2. Targeted Marketing Campaigns:
By utilizing the predictions of the model, marketing campaigns can be tailored to specific customer segments. Focus can be directed towards those customers with a high likelihood of making a purchase, optimizing marketing budget and resources.

3. Cost Efficiency:
Predicting customer behavior allows for the allocation of resources to the most promising leads, reducing the cost per conversion. This efficiency can lead to a more cost-effective marketing strategy.

4. Personalization:
The model enables a personalized approach to marketing by identifying individual customer preferences and behavior. This can result in more engaging and relevant marketing messages, increasing the likelihood of conversions.

5. Continuous Improvement:
The model can be periodically retrained with new data, allowing it to adapt to changing customer behavior patterns. This ensures that the predictions remain accurate over time and in evolving market conditions.

The model achieved an impressive accuracy of 98.42% during training, demonstrating its ability to accurately classify customers likely to make a purchase.

The predictions for new records further reinforce the model's effectiveness, with high probabilities associated with each prediction. Notably, the model identified customers with a strong likelihood of making a purchase, providing valuable insights for targeted marketing campaigns.

The benefits of implementing this tool in our marketing strategy are substantial. We can now make more informed decisions, allocate resources more efficiently, and personalize our approach to engage customers effectively. This not only enhances our marketing ROI but also positions us to stay ahead in a competitive market.